python/pyarrow/tests/test_parquet.py

```diff
@@ -2301,6 +2301,64 @@ def test_write_to_dataset_no_partitions(tempdir):
     _test_write_to_dataset_no_partitions(str(tempdir))
 
 
+@pytest.mark.pandas
+def test_write_to_dataset_with_partitions_and_custom_filenames(tempdir):
+    # ARROW-1400
+    output_df = pd.DataFrame({'group1': list('aaabbbbccc'),
+                              'group2': list('eefeffgeee'),
+                              'num': list(range(10)),
+                              'nan': [pd.np.nan] * 10,
+                              'date': np.arange('2017-01-01', '2017-01-11',
+                                                dtype='datetime64[D]')})
+    cols = output_df.columns.tolist()
+    partition_by = ['group1', 'group2']
+    filename_separator = '-'
+    output_table = pa.Table.from_pandas(output_df)
+    path = str(tempdir)
+
+    def partition_filename_callback(keys):
+        return "{0}{sep}{1}".format(*keys, sep=filename_separator)
+
+    pq.write_to_dataset(output_table, path,
+                        partition_by, partition_filename_callback)
+
+    metadata_path = os.path.join(path, '_common_metadata')
+    pq.write_metadata(output_table.schema, metadata_path)
+
+    # ARROW-2891: Ensure the output_schema is preserved when writing a
+    # partitioned dataset
+    dataset = pq.ParquetDataset(path,
+                                validate_schema=True)
+
+    # ARROW-2209: Ensure the dataset schema also includes the partition columns
+    dataset_cols = set(dataset.schema.to_arrow_schema().names)
+    assert dataset_cols == set(output_table.schema.names)
```

Comment:
General comment: I don't think it is needed to repeat the (for this case unrelated) asserts that are already in the other tests. I would also remove the references to the other ARROW-xxxx issues that are not related to ARROW-3538. 
I would just focus on testing the additional aspect of the file names.

Reply:
Makes complete sense - was simply unsure if custom file names might cause any knock-on effects.